src/main/java/org/xstefank/model/PersistentList.java

```diff
@@ -30,32 +33,126 @@ public PersistentList(String dirName, String fileName) {
         this(new File(dirName, fileName));
     }
 
-    public void addUser(String username) {
-        if (list.contains(username)) {
-            return;
+    @Override
+    public String set(int index, String element) {
+        String oldValue = super.set(index, element);
+        writeListContentToFile();
+        return oldValue;
+    }
+
+    @Override
+    public boolean add(String s) {
+        if (contains(s)) {
+            return false;
         }
-        list.add(username);
         try (FileWriter fw = new FileWriter(file, true)) {
-            fw.write(username + "\n");
+            fw.write(s + "\n");
+            return super.add(s);
         } catch (IOException e) {
             throw new IllegalStateException("Cannot write user to file", e);
         }
     }
 
-    public boolean hasUsername(String  username) {
-        return list.contains(username);
+    @Override
+    public void add(int index, String element) {
+        super.add(index, element);
+        writeListContentToFile();
+    }
+
+    @Override
+    public String remove(int index) {
+        String element = super.remove(index);
+        writeListContentToFile();
+        return element;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        if (!contains(o)) {
+            return false;
+        }
+        remove(indexOf(o));
+        return true;
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        writeListContentToFile();
+    }
+
+    @Override
+    public boolean addAll(int index, Collection<? extends String> c) {
+        boolean changed = super.addAll(index, c);
+        writeListContentToFile();
+        return changed;
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends String> c) {
+        boolean changed = super.addAll(c);
+        try (FileWriter fw = new FileWriter(file, true)){
```

Comment:
Because if you use this method it just adds content of collection "c" after the last element in list. So It is safe to write it to file this way. I don't need to rewrite whole file.